testing/trino-testing/src/main/java/io/trino/testing/ThreadAssertions.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.testing;
+
+import io.trino.jvm.Threads;
+
+import java.lang.management.ManagementFactory;
+import java.lang.management.ThreadMXBean;
+import java.util.Arrays;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Consumer;
+
+import static io.airlift.concurrent.MoreFutures.getFutureValue;
+import static java.lang.Math.clamp;
+import static java.util.stream.Collectors.joining;
+
+public final class ThreadAssertions
+{
+    private ThreadAssertions() {}
+
+    private static final AtomicInteger sequence = new AtomicInteger();
+
+    public static <T extends AutoCloseable> void assertNoThreadLeaked(
+            ThrowingSupplier<T> resourceCreator,
+            Consumer<T> exerciseResource)
+            throws Exception
+    {
+        // warm up all statically initialized threads
+        try (T resource = resourceCreator.get()) {
+            exerciseResource.accept(resource);
+        }
+
+        ThreadMXBean threadMXBean = ManagementFactory.getThreadMXBean();
+        CompletableFuture<Void> testFuture = new CompletableFuture<>();
+        ThreadGroup threadGroup = new ThreadGroup("test-group-" + sequence.incrementAndGet());
+        new Thread(threadGroup, () -> {
+            try {
+                try (T resource = resourceCreator.get()) {
+                    exerciseResource.accept(resource);
+                }
+                Thread[] threads = new Thread[clamp(threadGroup.activeCount(), 10, 1000)];
```

Comment:
Why 10? 1000?

Reply:
This is brittle and will result in false positives, false negatives, and potentially flaky tests. How does it account for all the JVM platform threads or framework threads that are not under Trino's control?

Also, relying on an unspecified naming convention ("ForkJoinPool.commonPool-worker-xxx") is not a good idea, as there's no guarantee it won't change arbitrarily.

Reply:
You can do this to get the threads:
```
        List<Thread> list = Thread.getAllStackTraces().keySet().stream()
                .filter(thread -> threadGroup.parentOf(thread.getThreadGroup()))
                .toList();
```
It is slower because the JVM gets the stack traces also, but it is more accurate.

Reply:
> Why 10? 1000?

10 -- at least no-zero, so that the next call is authoritative on live threads.
1000 -- arbitrarily chosen cap so that we don't print too many threads, if too many threads stay alive

> This is brittle and will result in false positives, false negatives, and potentially flaky tests. How does it account for all the JVM platform threads or framework threads that are not under Trino's control?

This may be brittle and we may need to improve or remove the test.

false positives (successes) are acceptable from this test perspective.
false negatives (failures) are not acceptable and we may need to improve or remove the test.
potentially flaky tests -- thanks for pointing this out. You know how much i care about non-flaky tests and now i don't feel alone
account for JVM platform threads -- by inspecting only one ThraedGroup, created privately within the test.
framework threads -- by performing a warmup first (might not be sufficient if framework expires the threads, then we may eg need to add excludes for known non-leaks -- i had it initially)

Reply:
> You can do this to get the threads:

thanks @dain for the suggestion

Reply:
> > false positives (successes) are acceptable from this test perspective.
> 
> How so? Then what is the test verifying?

if the test doesn't find a leaked thread, than it's not worse than not having the test at all

if it finds some leaks, it's still useful -- and it fact it found quite a bunch already 

> In particular, the lambda passed to assertNoThreadLeaked may be leaking threads indirectly from thread pools that were created outside of the lambda (and not part of the threadgroup). These won't be detected.

that's a very good point.

The other imperfection of the test is that it doesn't cover virtual threads at all.
Virtual threads don't inherit the ThreadGroup they are created from.

Do you have a suggestion for an improvement?
Do you think this test brings negligible value?

Reply:
> if the test doesn't find a leaked thread, than it's not worse than not having the test at all

It is, because it provides a false sense of security that the test and the tested code are correct. It's valid for a very narrow set of use cases and comes with many caveats.

>  Do you think this test brings negligible value?

I think it's ok for very specific cases where we know exactly what we're testing. But I worry that people are going to blindly copy what they see and end up writing malformed tests. The fact that this is presented as generic infrastructure for testing for thread leakage makes it even more likely for that to happen.

Reply:
It's generic, because i wanted to test cover both query runner and PlanTester.
If i introduce the public methods that expect these types, would it help?
If not, what would you do?

Reply:
Do you want me to remove PlanTester within this PR? Not sure I know exactly what this entails.

Reply:
No, that would be a separate change. But that wouldn’t be a pre requisite for testing query runners alone.

Reply:
OK, so i kept the test with PlanTester and made it impossible to provide anything other than PlanTester or QueryRunner